react-components/src/components/RevealToolbar/LayersContainer/CadModelLayersContainer.tsx

```diff
@@ -0,0 +1,92 @@
+/*!
+ * Copyright 2023 Cognite AS
+ */
+
+import React, { type ReactElement, useState } from 'react';
+import { useReveal } from '../../RevealContainer/RevealContext';
+import { type CogniteCadModel } from '@cognite/reveal';
+import { Menu } from '@cognite/cogs.js';
+import { StyledCheckbox, StyledSubMenu, StyledMenu } from './elements';
+
+export const CadModelLayersContainer = (): ReactElement => {
+  const viewer = useReveal();
+  const cadModels = viewer.models.filter((model) => model.type === 'cad');
+
+  const [selectedCadModels, setSelectedCadModels] = useState<
+    Array<{ model: CogniteCadModel; isToggled: boolean }>
+  >(
+    cadModels.map((model) => ({
+      model: model as CogniteCadModel,
+      isToggled: (model as CogniteCadModel).visible
+    }))
+  );
+
+  const [allCadModelVisible, setAllCadModelVisible] = useState(true);
+  const [indeterminate, setIndeterminate] = useState<boolean>(false);
+
+  const handleCadModelVisibility = (model: CogniteCadModel): void => {
+    selectedCadModels.map((data) => {
+      if (data.model === model) {
+        data.isToggled = !data.isToggled;
+      }
+      return data;
+    });
+    model.visible = !model.visible;
+    viewer.requestRedraw();
+    setSelectedCadModels([...selectedCadModels]);
+    setIndeterminate(selectedCadModels.some((data) => !data.isToggled));
+    setAllCadModelVisible(!selectedCadModels.every((data) => !data.isToggled));
+  };
+
+  const handleAllCadModelsVisibility = (visible: boolean): void => {
+    selectedCadModels.forEach((data) => {
+      data.isToggled = visible;
+      data.model.visible = visible;
+    });
+    viewer.requestRedraw();
+    setAllCadModelVisible(visible);
+    setIndeterminate(false);
+    setSelectedCadModels([...selectedCadModels]);
+  };
+
+  const cadModelContent = (): React.JSX.Element => {
+    return (
+      <StyledSubMenu>
+        {selectedCadModels.map((data) => (
+          <Menu.Item
+            key={data.model.modelId}
+            hasCheckbox
+            checkboxProps={{
+              checked: data.isToggled,
+              onChange: (e: { stopPropagation: () => void }) => {
+                e.stopPropagation();
+                handleCadModelVisibility(data.model);
+              }
+            }}>
+            {data.model.modelId}
+          </Menu.Item>
+        ))}
+      </StyledSubMenu>
+    );
+  };
+
+  return (
+    <>
+      {selectedCadModels.length > 0 ? (
+        <StyledMenu>
+          <StyledCheckbox
+            checked={allCadModelVisible}
+            indeterminate={indeterminate}
+            onChange={(e, c) => {
+              e.stopPropagation();
+              handleAllCadModelsVisibility(c as boolean);
+            }}
+          />
+          <Menu.Submenu content={cadModelContent()}>CAD models</Menu.Submenu>
```

Comment:
Can be removed with the changes above. Then the content of menu item should like something like this: 
```
<Flex direction='row' justifyContent='space-between'>
    <Checkbox
        checked={allCadModelVisible}
        onChange={(e, c) => {
            e.stopPropagation();
            handleAllCadModelsVisibility(c as boolean);
        }}
    />
    <div>
        Cad models
    </div>
</Flex>
```

Reply:
The same should work for point clouds/360 images.

Reply:
Also, we agreed to add counter for number of loaded models on the right side of the menu item as it is stated in the design.